File: src/connectors/whois_native.js
import { BaseConnector } from './base.js';
import net from 'net';

export class WhoisNativeConnector extends BaseConnector {
    constructor(config, state, telemetry) {
        super(config, state, telemetry);
        this.name = 'WHOIS-Native';
    }

    async run(investigationId, phaseId, inputType, inputValue) {
        const runId = this.state.recordConnectorRun(investigationId, phaseId, this.name, inputType, inputValue);
        this.telemetry.connectorStart(investigationId, this.name, phaseId, { type: inputType, value: inputValue });

        try {
            const raw = await this._whoisQuery(inputValue);
            const entities = this._parseWhois(raw, inputValue);

            let newCount = 0;
            for (const entity of entities) {
                const added = this.state.addEntity(investigationId, entity.type, entity.data, entity.confidence || 0.5);
                if (added.new) newCount++;
                this.telemetry.entityDiscovered(investigationId, entity.type, added.new, this.name);
                if (entity.asKnown) {
                    this.state.addKnown(investigationId, entity.asKnown.type, entity.asKnown.value, this.name, entity.confidence || 0.5);
                }
            }

            this.state.completeConnectorRun(runId, 'completed', newCount, raw?.slice(0, 5000), null, 0);
            this.telemetry.connectorEnd(investigationId, this.name, phaseId, {
                status: 'completed', entitiesFound: newCount, input: { type: inputType, value: inputValue }
            });
            return { status: 'completed', entities, newCount };
        } catch (e) {
            this.state.completeConnectorRun(runId, 'failed', 0, null, e.message, 1);
            this.telemetry.connectorEnd(investigationId, this.name, phaseId, {
                status: 'failed', entitiesFound: 0, input: { type: inputType, value: inputValue }
            });
            return { status: 'failed', error: e.message };
        }
    }

    _parseWhois(raw, domain) {
        const entities = [];
        const fields = {};

        for (const line of raw.split('\n')) {
            const match = line.match(/^\s*([\w\s/]+):\s*(.+)$/);
            if (match) {
                const key = match[1].trim().toLowerCase();
                const val = match[2].trim();
                if (!fields[key]) fields[key] = val;
            }
        }

        entities.push({
            type: 'domain',
            data: {
                name: domain,
                registrar: fields['registrar'] || fields['registrar name'],
                created_at: fields['creation date'] || fields['created'] || fields['registered'],
                expires_at: fields['registry expiry date'] || fields['expiration date'] || fields['expires'],
                updated_at: fields['updated date'] || fields['last updated'],
                nameservers: this._extractNameservers(raw),
                status: fields['domain status'],
                dnssec: fields['dnssec']
            },
            confidence: 0.95
        });

        const emails = raw.match(/[\w.+-]+@[\w-]+\.[\w.-]+/g);
        if (emails) {
            const seen = new Set();
            for (const email of emails) {
                if (!seen.has(email.toLowerCase()) && !email.includes('abuse')) {
                    seen.add(email.toLowerCase());
                    entities.push({
                        type: 'email',
                        data: { address: email, source: 'whois' },
                        confidence: 0.7,
                        asKnown: { type: 'email', value: email }
                    });
                }
            }
        }

        const ns = this._extractNameservers(raw);
        for (const nameserver of ns) {
            entities.push({
                type: 'subdomain',
                data: { name: nameserver, source: 'whois_ns' },
                confidence: 0.9
            });
        }

        const org = fields['registrant organization'] || fields['org-name'] || fields['organization'];
        if (org && org !== 'REDACTED FOR PRIVACY' && !org.includes('Privacy') && !org.includes('Proxy')) {
            entities.push({
                type: 'organization',
                data: { name: org, source: 'whois_registrant' },
                confidence: 0.7
            });
        }

        const registrantName = fields['registrant name'];
        if (registrantName && registrantName !== 'REDACTED FOR PRIVACY' && !registrantName.includes('Privacy')) {
            entities.push({
                type: 'person',
                data: { name: registrantName, source: 'whois_registrant' },
                confidence: 0.6
            });
        }

        return entities;
    }

    _extractNameservers(raw) {
        const ns = [];
        const regex = /name\s*server:\s*(\S+)/gi;
        for (const match of raw.matchAll(regex)) {
            ns.push(match[1].toLowerCase().replace(/\.$/, ''));
        }
        return [...new Set(ns)];
    }

    _whoisQuery(domain, server = 'whois.verisign-grs.com') {
        return new Promise((resolve, reject) => {
            const socket = net.createConnection(43, server, () => {
                socket.write(domain + '\r\n');
            });
            let data = '';
            socket.on('data', chunk => { data += chunk.toString(); });
            socket.on('end', () => {
                const referral = data.match(/Registrar WHOIS Server:\s*(\S+)/i);
                if (referral && referral[1] !== server) {
                    this._whoisQuery(domain, referral[1]).then(resolve).catch(() => resolve(data));
                } else {
                    resolve(data);
                }
            });
            socket.on('error', reject);
            socket.setTimeout(15000, () => { socket.destroy(); reject(new Error('whois timeout')); });
        });
    }

    async healthCheck() { return { ok: true, version: 'native' }; }
}

File: cyclops/connectors/whois_native.py
import asyncio
import socket
from cyclops.connectors.base import BaseConnector


class WhoisNativeConnector(BaseConnector):
    def __init__(self, config, state, telemetry):
        super().__init__(config, state, telemetry)
        self.name = "WHOIS-Native"

    async def run(self, investigation_id, phase_id, input_type, input_value):
        run_id = self.state.record_connector_run(investigation_id, phase_id, self.name, input_type, input_value)
        self.telemetry.connector_start(investigation_id, self.name, phase_id, {"type": input_type, "value": input_value})

        try:
            loop = asyncio.get_event_loop()
            raw = await loop.run_in_executor(None, self._whois_query, input_value)
            entities = self._parse_whois(raw, input_value)

            new_count = 0
            for entity in entities:
                added = self.state.add_entity(investigation_id, entity["type"], entity["data"], entity.get("confidence", 0.5))
                if added["new"]:
                    new_count += 1
                self.telemetry.entity_discovered(investigation_id, entity["type"], added["new"], self.name)

            self.state.complete_connector_run(run_id, "completed", new_count)
            self.telemetry.connector_end(investigation_id, self.name, phase_id, {"status": "completed", "entitiesFound": new_count, "input": {"type": input_type, "value": input_value}})
            return {"status": "completed", "entities": entities, "newCount": new_count}
        except Exception as e:
            self.state.complete_connector_run(run_id, "failed", 0, error=str(e))
            self.telemetry.connector_end(investigation_id, self.name, phase_id, {"status": "failed", "input": {"type": input_type, "value": input_value}})
            return {"status": "failed", "error": str(e)}

    def _whois_query(self, domain):
        sock = socket.socket(socket.AF_INET, socket.SOCK_STREAM)
        sock.settimeout(10)
        try:
            sock.connect(("whois.verisign-grs.com", 43))
            sock.send(f"={domain}\r\n".encode())
            data = b""
            while True:
                chunk = sock.recv(4096)
                if not chunk:
                    break
                data += chunk
            first_pass = data.decode("utf-8", errors="replace")
            refer = None
            for line in first_pass.split("\n"):
                if line.strip().lower().startswith("whois server:"):
                    refer = line.split(":", 1)[1].strip()
                    break
            if refer:
                sock2 = socket.socket(socket.AF_INET, socket.SOCK_STREAM)
                sock2.settimeout(10)
                try:
                    sock2.connect((refer, 43))
                    sock2.send(f"{domain}\r\n".encode())
                    data2 = b""
                    while True:
                        chunk = sock2.recv(4096)
                        if not chunk:
                            break
                        data2 += chunk
                    return data2.decode("utf-8", errors="replace")
                finally:
                    sock2.close()
            return first_pass
        finally:
            sock.close()

    def _parse_whois(self, raw, domain):
        entities = []
        info = {}
        for line in raw.split("\n"):
            if ":" not in line or line.strip().startswith("%") or line.strip().startswith("#"):
                continue
            key, _, value = line.partition(":")
            key = key.strip().lower()
            value = value.strip()
            if not value:
                continue
            if "registrar" in key and "registrar" not in info:
                info["registrar"] = value
            elif "creation" in key or "created" in key:
                info["created_at"] = value
            elif "expir" in key:
                info["expires_at"] = value
            elif "name server" in key or "nserver" in key:
                info.setdefault("nameservers", []).append(value.lower().rstrip("."))
            elif "registrant" in key and "name" in key:
                info["registrant_name"] = value
            elif "registrant" in key and "org" in key:
                info["registrant_org"] = value

        if info:
            entities.append({"type": "domain", "data": {"name": domain, "source": "whois", **info}, "confidence": 0.9})
        if info.get("registrant_org"):
            entities.append({"type": "organization", "data": {"name": info["registrant_org"], "source": "whois"}, "confidence": 0.7})
        for ns in info.get("nameservers", []):
            entities.append({"type": "subdomain", "data": {"name": ns, "source": "whois_ns"}, "confidence": 0.9})

        return entities

    async def health_check(self):
        return {"ok": True, "version": "native"}
